src/metabase/serialization/dump.clj

```diff
@@ -0,0 +1,294 @@
+                     (map #(-> % Integer/parseInt Collection :name (str "/collections")))
+                     (str/join "/")
+                     (format "%s/"))]
+    (str "/collections/collections/" parents (:name collection))))
+
+(defmethod fully-qualified-name (type Dashboard)
+  [dashboard]
+  (format "%s/dashboards/%s"
+          (or (some->> dashboard :collection_id Collection fully-qualified-name)
+              "/collections")
+          (:name dashboard)))
+
+(defmethod fully-qualified-name (type Pulse)
+  [pulse]
+  (format "%s/pulses/%s"
+          (or (some->> pulse :collection_id Collection fully-qualified-name)
+              "/collections")
+          (:name pulse)))
+
+(defmethod fully-qualified-name (type Card)
+  [card]
+  (format "%s/cards/%s"
+          (or (some->> card
+                       :dataset_query
+                       qp.util/query->source-card-id
+                       Card
+                       fully-qualified-name)
+              (some->> card
+                       :collection_id
+                       Collection
+                       fully-qualified-name)
+              "/collections")
+          (:name card)))
+
+(defmethod fully-qualified-name nil
+  [_]
+  nil)
+
+(def ^:private SegmentOrMetric
+  [(s/one (s/constrained su/KeywordOrString
+                         (comp #{:metric :segment} qp.util/normalize-token))
+          "head")
+   (s/cond-pre s/Int su/KeywordOrString)])
+
+(def ^{:arglists '([form])} entity-reference?
+  "Is given form an MBQL entity reference?"
+  (some-fn field-reference? (complement (s/checker SegmentOrMetric))))
+
+(defn- entity-reference->fully-qualified-name
+  [[op & args :as entity-reference]]
+  (case (qp.util/normalize-token op)
+    :metric        [op (fully-qualified-name (Metric (first args)))]
+    :segment       [op (fully-qualified-name (Segment (first args)))]
+    :field-id      [op (if (string? (first args))
+                         (first args)
+                         (fully-qualified-name (Field (first args))))]
+    :fk->          (into [op]
+                         (for [arg args]
+                           (if (number? arg)
+                             (Field arg)
+                             (entity-reference->fully-qualified-name arg))))
+    :field-literal entity-reference))
+
+(defn- humanize-entity-references
+  [entity]
+  (walk/postwalk
+   (fn [form]
+     (cond
+       (entity-reference? form)
+       (entity-reference->fully-qualified-name form)
+
+       (map? form)
+       (let [id->fully-qualified-name (fn [entity-id model]
+                                        (if (string? entity-id)
+                                          entity-id
+                                          (fully-qualified-name (model entity-id))))]
+         (-> form
+             (u/update-when :database (fn [db]
+                                        (if (= db -1337)
+                                          "database/virtual"
```

Comment:
ok, what happens if someone names their database `virtual`?

Reply:
Also from a code hygiene standpoint instead of using the magic number it would be preferable to use the var wherever that lives these days. Ideally if we decide we want to change the magic ID to `-1337` to `-2000` for example we should only need to change one place instead of having to track down every reference to it; `(= db database/virtual-id)` is also a lot clearer than `(= db -1337)` from a readability standpoint